components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinksUseCases.kt

```diff
@@ -76,74 +76,81 @@ class AppLinksUseCases(
         private val includeInstallAppFallback: Boolean = false
     ) {
         operator fun invoke(url: String): AppLinkRedirect {
-            val intents = createBrowsableIntents(url)
-            val (appIntent, resolveInfos) = if (includeHttpAppLinks) {
-                intents.asSequence()
-                    .map { it to getNonBrowserActivities(it) }
-                    .firstOrNull { it.second.isNotEmpty() }
-                    ?.let {
-                        // The user may have decided to keep opening this type of link in this browser.
-                        if (ignoreDefaultBrowser &&
-                            findDefaultActivity(it.first)?.activityInfo?.packageName == context.packageName) {
-                            // in which case, this isn't an app intent anymore.
-                            null
-                        } else {
-                            it
-                        }
-                    }
-            } else {
-                intents.asSequence()
-                    .filter { it.data?.isHttpOrHttps != true }
-                    .map { it to getNonBrowserActivities(it) }
-                    .firstOrNull { it.second.isNotEmpty() }
-            } ?: null to null
-
-            val webUrls = intents.mapNotNull {
-                if (it.data?.isHttpOrHttps == true) it.dataString else null
+            val redirectData = createBrowsableIntents(url)
+            val isAppIntentHttpOrHttps = when (redirectData.appIntent?.data?.isHttpOrHttps) {
+                true -> true
+                else -> false
+            }
+            var appIntent = redirectData.appIntent
+            if (redirectData.resolveInfo == null) {
+                appIntent = null
+            }
+
+            if (includeHttpAppLinks && (ignoreDefaultBrowser || isDefaultBrowser(redirectData.appIntent))) {
+                appIntent = null
+            }
+
+            if (!includeHttpAppLinks && isAppIntentHttpOrHttps) {
+                appIntent = null
```

Comment:
Can we rewrite this as a `when`?

```kotlin
val appIntent = when {
    redirectData.appIntent == null -> null
    redirectData.resolveInfo == null -> null
    includeHttpAppLinks && (ignoreDefaultBrowser || isDefaultBrowser(redirectData.appIntent)) -> null
    !includeHttpAppLinks && isAppIntentHttpOrHttps -> null
    else -> redirectData.appIntent
}
```

Reply:
Didn't think about writing it this way.  Much cleaner, I'll update.  Thanks,